### PR TITLE
Support cache

### LINKED ISSUE #1652: [Proposal] Support cache
Support using `actions-cache@v3`.

There's a basic implementation for cache in act_runner(the CICD runner of Gitea, based on act). See https://gitea.com/gitea/act_runner/pulls/25 .  And it seems to be working well so far. 

It's an independent module, so I think it's not difficult to port it to act. However, I'm not sure if act really needs it, so I need some feedback from the maintainers before I do anything.

Related to https://github.com/nektos/act/issues/1513 .

### PR DESCRIPTION
**Pull request by wolfogre**
Close #1652 

Ported from [act runner](https://gitea.com/gitea/act_runner), but it's quite different now:

- Use bolt db instead of sqlite.
- Use httprouter instead of chi.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -609,6 +610,13 @@ func newRunCommand(ctx context.Context, input *Input) func(*cobra.Command, []str
 
 		cancel := artifacts.Serve(ctx, input.artifactServerPath, input.artifactServerAddr, input.artifactServerPort)
 
+		// TODO: provide a way to configure the cache handler via cli
+		cacheHandler, err := artifactcache.StartHandler("", "", 0, common.Logger(ctx))
+		if err != nil {
+			return err
+		}
+		envs["ACTIONS_CACHE_URL"] = cacheHandler.ExternalURL() + "/"
```
Comment (KnisterPeter) [inline cmd/root.go:618]: Can we make this optional? We have an external cache which can run across a cluster of nodes and uses it's own external backend.
Having this here will probably interfer with it, right?

I think the cache handling should be opt-in similar to the asset-server.

**Review by KnisterPeter — CHANGES_REQUESTED** [review]

File: pkg/artifactcache/model.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	Version        string `json:"version" boltholdIndex:"Version"`
+	KeyVersionHash string `json:"keyVersionHash" boltholdUnique:"KeyVersionHash"`
+	Size           int64  `json:"cacheSize"`
+	Complete       bool   `json:"complete"`
+	UsedAt         int64  `json:"usedAt" boltholdIndex:"UsedAt"`
+	CreatedAt      int64  `json:"createdAt" boltholdIndex:"CreatedAt"`
+}
+
+func (c *Cache) FillKeyVersionHash() {
+	c.KeyVersionHash = fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%s:%s", c.Key, c.Version))))
```
[the remark this answers is not in the document]
Comment (wolfogre) [reply pkg/artifactcache/model.go:37]: Fixed in 5c1c2115b4cbe44cea227910820301e3e3da9db9 , just intercept at the entrance.

File: cmd/root.go [hunk 3]
```diff
@@ -609,6 +610,13 @@ func newRunCommand(ctx context.Context, input *Input) func(*cobra.Command, []str
 
 		cancel := artifacts.Serve(ctx, input.artifactServerPath, input.artifactServerAddr, input.artifactServerPort)
 
+		// TODO: provide a way to configure the cache handler via cli
+		cacheHandler, err := artifactcache.StartHandler("", "", 0, common.Logger(ctx))
```
Comment (KnisterPeter) [inline cmd/root.go:614]: Using '' as the default `dir` value will implicitly store probably gigabytes of data in the users homedir without users noticing.
Should this be better a custom temp folder which is cleanup by the OS?

Don't get me wrong, it should be configurable, but by default we should be more conservative with users diskspace.

Comment (wolfogre) [reply cmd/root.go]: '' means default path (`$HOME/.cache/actcache`) for `artifactcache. StartHandler`, and I made it configurable now:

image

So default value could be `$XDG_CACHE_HOME/actcache` or `$HOME/.cache/actcache`, I think it's fair enought, right?

And don't worry, artifactcache will clean the old cache even the OS does't do it.

File: cmd/root.go (hunk 1 above)
Comment (KnisterPeter) [reply cmd/root.go]: I think it's fine, as we are setting all required env vars.

**Review by KnisterPeter — APPROVED** [review]
Would be fine to only check if env vars are set but it's okay to have an extra flag. Any way would work for us.

**Pull request merged**